homeassistant/components/smartthings/cover.py

```diff
@@ -0,0 +1,146 @@
+        self._device_class = None
+        self._state = None
+        self._state_attrs = None
+        self._supported_features = SUPPORT_OPEN | SUPPORT_CLOSE
+        if Capability.switch_level in device.capabilities:
+            self._supported_features |= SUPPORT_SET_POSITION
+
+    async def async_close_cover(self, **kwargs):
+        """Close cover."""
+        # Same command for all 3 supported capabilities
+        await self._device.close(set_status=True)
+        # State is set optimistically in the commands above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state(True)
+
+    async def async_open_cover(self, **kwargs):
+        """Open the cover."""
+        # Same for all capability types
+        await self._device.open(set_status=True)
+        # State is set optimistically in the commands above, therefore update
+        # the entity state ahead of receiving the confirming push updates
+        self.async_schedule_update_ha_state(True)
+
+    async def async_set_cover_position(self, **kwargs):
+        """Move the cover to a specific position."""
+        position = kwargs.get(ATTR_POSITION)
+        if not self._supported_features & SUPPORT_SET_POSITION \
+                or not position:
+            return
+        # Do not set_status=True as device will report progress.
+        await self._device.set_level(position, 0)
+
+    async def async_update(self):
+        """Update the attrs of the cover."""
+        from pysmartthings import Attribute, Capability
+
+        value = None
+        if Capability.door_control in self._device.capabilities:
+            self._device_class = DEVICE_CLASS_DOOR
+            value = self._device.status.door
+        elif Capability.window_shade in self._device.capabilities:
+            self._device_class = DEVICE_CLASS_SHADE
+            value = self._device.status.window_shade
+        elif Capability.garage_door_control in self._device.capabilities:
+            self._device_class = DEVICE_CLASS_GARAGE
+            value = self._device.status.door
+
+        self._state = VALUE_TO_STATE.get(value)
+
+        self._state_attrs = {}
+        battery = self._device.status.attributes[Attribute.battery].value
+        if battery is not None:
+            self._state_attrs[ATTR_BATTERY_LEVEL] = battery
+
+    @property
+    def current_cover_position(self):
+        """Return current position of cover."""
+        return self._device.status.level
+
+    @property
+    def device_class(self):
+        """Define this cover as a garage door."""
+        return self._device_class
+
+    @property
+    def device_state_attributes(self):
+        """Get additional state attributes."""
+        return self._state_attrs
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return self._supported_features
+
+    @property
+    def state(self) -> str:
```

Comment:
We shouldn't overwrite this property. The cover base entity class is responsible for this. Implement `is_opening`, `is_closing` and `is_closed`.

Reply:
Is that a hard rule?  Overriding it resulted in less code that is more maintainable.  Otherwise, I can't map the value directly (because it could be broken across the 3 properties), and there's some confusing/unexpected behavior of needing to return `None` from `is_closed` to get `Unknown`.

Reply:
Yes, it's a hard rule.

Reply:
OK!